File: mail/src/main/java/com/example/mail/MailService.java
package com.example.mail;

import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@Service
public class MailService {
    private final Map<Long, Mail> mails = new HashMap<>();


    public Mail saveMail(Mail mail) {
        Long uuid = generateUUID();
        mail.setUuid(uuid);
        mails.put(uuid, mail);
        return mail;
    }

    private Long generateUUID() {
        return (long) (mails.size() + 1);
    }

    public Mail updateMail(Long uuid, Mail mail) {
        if (!mails.containsKey(uuid)) {
            throw new IllegalArgumentException("No such mail!");
        }
        mail.setUuid(uuid);
        mails.put(uuid, mail);
        return mail;
    }

    public Mail getMail(Long uuid) {
        return Optional.ofNullable(mails.get(uuid)).orElseThrow(() -> new IllegalArgumentException("No such mail!"));
    }

    public Collection<Mail> getAllMail() {
        return mails.values();
    }
}

File: mail/src/main/kotlin/com/example/mail/MailService.kt
package com.example.mail

import org.springframework.stereotype.Service
import java.util.Optional

@Service
class MailService {
    private var mails = mutableMapOf<Long, Mail>()

    fun saveMail(mail: Mail): Mail {
        val uuid = generateUUID()
        mail.uuid = uuid
        mails[uuid] = mail
        return mail
    }

    private fun generateUUID() = (mails.size + 1).toLong()

    fun getMail(uuid: Long): Mail {
        val mail = Optional.ofNullable(mails[uuid]).orElseThrow {
            throw IllegalArgumentException("No such mail!")
        }
        return mail
    }

    fun getAllMail() = mails.values.filter { mail -> !mail.delivered }

    fun saveAll(vararg args: Mail) {
        args.forEach { mail ->
            saveMail(mail)
        }
    }

    fun updateMail(uuid: Long, mail: Mail): Mail {
        if (!mails.containsKey(uuid)) {
            throw IllegalArgumentException("No such mail!")
        }
        mail.uuid = uuid
        mails[uuid] = mail
        return mail
    }
}

